### ISSUE TITLE
Surface API Usability roughness

### ISSUE THREAD

**Issue by MingweiSamuel** [hydro-project/hydroflow#42]
- [ ] Handoffs require specifying a specific `CanReceive<T>` type. Previously we used this to allow submitting multiple types, but it adds roughness since we have to specify the `T` when using the surface API
- [ ] Forgetting to call `.flush()` on inputs is an easy footgun
- [ ] FlatMap needed since handoffs return their inner Vec instead of the items inside the vec. Also need a dedicated `.flatten()` instead of `.flat_map(std::convert::identity)`

**Comment by justinj** [comment]
Will add stuff here as I play with it.
[ ] It would be nice if `add_subgraph` didn't require you to call `pivot`, and could automatically convert a purely push or pull graph into a pivot one. Maybe just using `Into<Pivot>` or something would be enough.

**Comment by MingweiSamuel** [comment]
Oh yeah, forgetting to call `pivot()` was one.

Also related, the compiler tries to turn things into `Iterator` when it shouldn't if you call map/filter/for_each/etc without having the right trait bounds or imports.

**Comment by justinj** [comment]
* [ ] `flatten` as an alias for `.flat_map(std::convert::identity)` would be nice

**Comment by justinj** [comment]
* [ ] Currently you can pipe directly to a handoff via `reverse`, but it might be nice to have a more intuitive alias for that like `pipe`

**Comment by MingweiSamuel** [comment]
> * Currently you can pipe directly to a handoff via `reverse`, but it might be nice to have a more intuitive alias for that like `pipe`

Yeah, also maybe just rename `reverse` to something more user-facing

**Comment by MingweiSamuel** [comment]
- [ ] Needs to be a way to add an arbitrary source

**Comment by MingweiSamuel** [comment]
Also HydroflowBuilder is missing ::new() :P

**Comment by justinj** [comment]
I'd like a `constant`, just like, an operator that emits some fixed set of values:
```
HydroflowBuilder::constant([1, 2, 3])...
```
or something. I started to implement such a thing but I wonder if it would be better to have a general escape hatch like we had with `add_subgraph` before, so I could just insert some code to run as a sink/source/whatever? Then something like `constant` could be implemented in terms of that?

**Comment by justinj** [comment]
Another operator I'd like is `inspect`, which is basically `for_each` but also passes through its data. I think this can probably just be implemented in terms of `map`?

**Comment by justinj** [comment]
Is there a good reason Handoffs shouldn't auto `flatten`? I like having access to the batches but it has been somewhat error-prone for us so far.

**Comment by justinj** [comment]
Another important thing we need is a push-based version of the joins, we can work around it for now by putting a handoff boundary but it would be nice to avoid that.

**Comment by MingweiSamuel** [comment]
> Is there a good reason Handoffs shouldn't auto `flatten`? I like having access to the batches but it has been somewhat error-prone for us so far.

There just isn't really an iterator abstraction built into the handoffs since we always have had the intention to use them with lattices eventually.

> Another important thing we need is a push-based version of the joins, we can work around it for now by putting a handoff boundary but it would be nice to avoid that.

I'm not sure how the shared ownership of the output would work?

**Comment by justinj** [comment]
> There just isn't really an iterator abstraction built into the handoffs since we always have had the intention to use them with lattices eventually.

Good point!

> I'm not sure how the shared ownership of the output would work?

We could have one input be push and one be pull, I guess?

**Comment by justinj** [comment]
> There just isn't really an iterator abstraction built into the handoffs since we always have had the intention to use them with lattices eventually.

I was thinking about this more, maybe the right approach is to not allow you to use them directly as `PullSurfaces`, and require you to specify how you want the data out of them, like `iter_elements`, `iter_batches`, or something?

**Comment by justinj** [comment]
> We could have one input be push and one be pull, I guess?

Never mind, don't know what I was thinking here :/ not sure how to make such a thing work.

**Comment by justinj** [comment]
* It's pretty tricky to pull groups of operators out into a function or something, there's a lot of types to wrangle.
* It would be nice to have a `tee` that, rather than taking the two outward branches as arguments, takes _one_ of them as an argument, and then allows me to keep chaining on the `tee` itself.

**Comment by jhellerstein** [comment]
> * It's pretty tricky to pull groups of operators out into a function or something, there's a lot of types to wrangle.
> * It would be nice to have a `tee` that, rather than taking the two outward branches as arguments, takes _one_ of them as an argument, and then allows me to keep chaining on the `tee` itself.

Seconded. It's weird to write the tee outputs (`start_tee()`) first and the the tee input second. Feels backwards.

**Comment by justinj** [comment]
I think we're talking about different things, to be clear, I want:
```
v.tee(
  start_tee().foo.bar.baz,
).foo2.bar2.baz2
```
rather than
```
v.tee(
  start_tee().foo.bar.baz,
  start_tee().foo2.bar2.baz2,
)
```
but I don't mind the start_tee() stuff fundamentally.
